Dockerfile

```diff
@@ -2,8 +2,6 @@ FROM node:6.13.0-alpine
 
 WORKDIR /usr/app
 
-RUN npm install knex -g
```

Comment:
Would removing this not cause all db operations to fail?

Reply:
I think it's okay because we now only call knex using an npm script defined in package.json.

I can double check this.

Reply:
Yep this works for me, it would be good if you could try out using Docker Compose and let me know if you think it's better than installing everything locally.

Reply:
It works for me too, and I think it is a lot better than installing everything locally. I think the views need to be updated though to use the new files directory for user files.

Reply:
So I definitely misunderstood the problem in my previous comment, but I think I know what's happening now. Firstly, for me the volumes weren't actually being created for some reason. I fixed that by adding the following lines at the end of `docker-compose.yml`: 
```
volumes:
  app:
  postgres:
```
This turns them into named volumes I think, and to be honest I don't know why it changed anything.

Then since the `files/...` source directories located on the host were empty, the contents of the target directories, e.g. `/usr/app/public/...` were deleted when they are mounted so I had no avatars and such. Any files which I did put in the source directories then were appearing in the container and the application though.

I'd just like to add that these problems I'm experiencing may purely be from my setup, so I would trust your own testing more than mine!

Reply:
Weird that adding those named volumes caused the other volumes to be created as they seem unrelated? When I run it without the `files` and `postgres-data` directories it creates them but they belong to the root user.

Yeah the `/usr/app/public/...` directories should be loaded from the `files` directory. This allows the avatars, manifestos etc to persist whenever the container is destroyed. So this seems to be working as expected for you?

I'm not sure this is quite ready yet because if you make any changes to the code you have to rebuild the images...

Reply:
I've just double checked - for *me*, it does create all the directories, but using the `docker volume ls` command shows that no new volumes have actually been created. Making the volumes named is what fixes this for me. 

After doing this, it all works as expected for me.

Reply:
There's something dodgy going on with the files for me, not sure what yet but I can't get any avatars to appear at all and if I upload a file in the admin section it won't appear unless I close and reopen the browser.
I tried adding the volumes: lines of code but, as far as I could tell, I don't think that they seemed to make any difference to mine

Reply:
If you put some files (just copy the ones currently in `public/images/avatars`) into the `files/avatars` directory then they should appear on the website, in theory.

Reply:
I've realised that the avatars not appearing might just be a problem with internet explorer as the actual live website is also not showing avatars for me at the moment on internet explorer (but works fine in chrome). 

For some reason I can only run a local version properly on internet explorer, I've tried to use chrome or edge but can't login to it using them for some reason, but this isn't to do with docker compose as I've had the problem since I first installed it and was running with npm commands (Think it was something weird to do with file permission and I never really worked out why)

Reply:
Sounds like a separate issue. Hopefully if we can get docker working this should solve any permissions problems because it will ask be containerised. Needs more work though because if you make any changes to the actual code you have to rebuild the whole docker image to see these changes.

Reply:
Admittedly my understanding is limited, but is this actually avoidable? It seems the `docker commit` command is the easiest way to incorporate changes, but it still creates a completely new image.

Reply:
Yep, it seems like the permissions problems are solved as no errors come up in the logs when using it in docker, not sure what is causing the login issue for me on chrome then.

I was using docker-compose build to make any changes in the code update, is docker commit better? I tried looking it up but couldn't work out how to use it. 
I think it seems like it's theoretically possible to be avoidable as the answer on this question https://stackoverflow.com/questions/24272535/rebuild-container-after-each-change sound like he has changes updating without doing anything, but I don't really know what he's doing and it's probably quite a different set up to this so I don't know if any of that actually applies.

Reply:
Okay I'm going to try and put in place the kind of setup described in that link. To do so I've moved all the source code in to one directory in this pull request:
https://github.com/ben-willis/grey-jcr/pull/36

If either/both of you could take a look at that one and make sure I haven't missed anything obvious that would be great. In the mean time I'll close this request and do a bit more work to iron out some of the issues I've cause.

Sorry its a bit of a mess but thank you both for your help!